File: packages/core/src/research_engine/domain/works_files.py
"""Work files — created works as markdown with machine-checked citations.

Phase 0 holds a work as a file, not rows: the front matter carries structured
span citations (`document_id` plus offsets into canonical text) and the body
carries `[^cN]` markers. Later steps move this into `authored.*`; the field
names here are exactly the columns there, so the port is a transcription.
"""

from __future__ import annotations

import re
from datetime import date
from enum import StrEnum
from typing import Any
from uuid import UUID  # noqa: TC003 - pydantic needs it at runtime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class Intent(StrEnum):
    """Why a citation is here — the citation's own vocabulary."""

    QUOTATION = "quotation"
    TRANSLATION = "translation"
    SUPPORT = "support"
    CONTRAST = "contrast"
    BACKGROUND = "background"
    DEFINITION = "definition"
    SOURCE = "source"
    SEE_ALSO = "see_also"


class Role(StrEnum):
    """What a citation does for a claim — the ledger's vocabulary.

    Present only when the citation also backs one of the `claims:` refs.
    """

    ASSERTS = "asserts"
    SUPPORTS = "supports"
    REBUTS = "rebuts"
    CONTEXT = "context"


class WorkType(StrEnum):
    TRANSLATION = "translation"
    ESSAY = "essay"
    DOSSIER = "dossier"
    SCRIPT = "script"
    OUTLINE = "outline"


class WorkStatus(StrEnum):
    DRAFT = "draft"
    REVIEW = "review"
    PUBLISHED = "published"


#: Citation ids are `[^cN]` handles: `c` followed by digits, nothing else.
ENTRY_ID_RE = re.compile(r"^c[0-9]+$")


class CitationEntry(BaseModel):
    """One structured span citation from a work's front matter."""

    model_config = ConfigDict(extra="forbid")

    id: str
    intent: Intent
    role: Role | None = None
    document_id: UUID
    char_start: int
    char_end: int
    quoted_text: str
    edition: str | None = None
    edition_key: str | None = None
    locator: dict[str, Any] = Field(default_factory=dict)

    @field_validator("id")
    @classmethod
    def _id_is_a_handle(cls, value: str) -> str:
        if not ENTRY_ID_RE.match(value):
            raise ValueError(f"id must match ^c[0-9]+$, got {value!r}")
        return value

    @field_validator("char_start")
    @classmethod
    def _start_is_an_offset(cls, value: int) -> int:
        if value < 0:
            raise ValueError(f"char_start must be non-negative, got {value}")
        return value

    @field_validator("quoted_text")
    @classmethod
    def _quote_is_non_empty(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("quoted_text must be non-empty")
        return value

    @model_validator(mode="after")
    def _span_is_well_formed(self) -> CitationEntry:
        if self.char_end <= self.char_start:
            raise ValueError(
                f"char_end ({self.char_end}) must exceed "
                f"char_start ({self.char_start})"
            )
        return self


class WorkFrontMatter(BaseModel):
    """The validated header of a work file. Entries that failed validation are
    not here — they are in `WorkFile.entry_errors`, so one bad entry does not
    hide the others."""

    model_config = ConfigDict(extra="forbid")

    work: str
    title: str
    type: WorkType
    status: WorkStatus
    created: date
    claims: list[str] = Field(default_factory=list)
    citations: list[CitationEntry] = Field(default_factory=list)


class EntryError(BaseModel):
    """One front-matter entry that failed validation, with the reason.

    Not an exception: the file still parses, the other entries still verify,
    and `work_verify` reports this as `AUTH_ENTRY_INVALID`.
    """

    citation_id: str | None = None
    message: str


class WorkFile(BaseModel):
    """A parsed work file: header models, body text, and marker handles."""

    work_path: str
    front_matter: WorkFrontMatter
    #: sha256 hex of the raw YAML block bytes, fences excluded. Drift detection
    #: for the Step 4 mirror compares this, not a re-serialization.
    front_matter_sha: str
    body: str
    #: Every `[^cN]` in the body, in order, excluding footnote definition lines.
    markers: list[str] = Field(default_factory=list)
    entry_errors: list[EntryError] = Field(default_factory=list)

File: crates/marginalia-types/src/works_files.rs
//! Work files mirroring `domain/works_files.py`: markdown works with
//! machine-checked citations.

use chrono::NaiveDate;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::errors::{Error, Result};

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Intent {
    Quotation,
    Translation,
    Support,
    Contrast,
    Background,
    Definition,
    #[default]
    Source,
    SeeAlso,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Role {
    Asserts,
    Supports,
    Rebuts,
    Context,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum WorkType {
    Translation,
    Essay,
    Dossier,
    Script,
    Outline,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum WorkFileStatus {
    Draft,
    Review,
    Published,
}

/// Citation ids are `[^cN]` handles: `c` followed by digits, nothing else.
pub fn is_entry_id(value: &str) -> bool {
    let mut chars = value.chars();
    if chars.next() != Some('c') {
        return false;
    }
    let rest = chars.as_str();
    !rest.is_empty() && rest.bytes().all(|b| b.is_ascii_digit())
}

/// One structured span citation from a work's front matter.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct CitationEntry {
    pub id: String,
    pub intent: Intent,
    #[serde(default)]
    pub role: Option<Role>,
    pub document_id: Uuid,
    pub char_start: i64,
    pub char_end: i64,
    pub quoted_text: String,
    #[serde(default)]
    pub edition: Option<String>,
    #[serde(default)]
    pub edition_key: Option<String>,
    #[serde(default)]
    pub locator: Map<String, Value>,
}

impl CitationEntry {
    /// The `_id_is_a_handle`, `_start_is_an_offset`, `_quote_is_non_empty`,
    /// and `_span_is_well_formed` validators.
    pub fn validate(&self) -> Result<()> {
        if !is_entry_id(&self.id) {
            return Err(Error::Validation(format!(
                "id must match ^c[0-9]+$, got {:?}",
                self.id
            )));
        }
        if self.char_start < 0 {
            return Err(Error::Validation(format!(
                "char_start must be non-negative, got {}",
                self.char_start
            )));
        }
        if self.quoted_text.trim().is_empty() {
            return Err(Error::Validation(
                "quoted_text must be non-empty".to_owned(),
            ));
        }
        if self.char_end <= self.char_start {
            return Err(Error::Validation(format!(
                "char_end ({}) must exceed char_start ({})",
                self.char_end, self.char_start
            )));
        }
        Ok(())
    }
}

/// The validated header of a work file. Entries that failed validation live in
/// `WorkFile.entry_errors`, so one bad entry does not hide the others.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WorkFrontMatter {
    pub work: String,
    pub title: String,
    #[serde(rename = "type")]
    pub work_type: WorkType,
    pub status: WorkFileStatus,
    pub created: NaiveDate,
    #[serde(default)]
    pub claims: Vec<String>,
    #[serde(default)]
    pub citations: Vec<CitationEntry>,
}

/// One front-matter entry that failed validation, with the reason.
///
/// Not an error: the file still parses and `work_verify` reports it as
/// `AUTH_ENTRY_INVALID`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct EntryError {
    #[serde(default)]
    pub citation_id: Option<String>,
    pub message: String,
}

/// A parsed work file: header models, body text, and marker handles.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WorkFile {
    pub work_path: String,
    pub front_matter: WorkFrontMatter,
    /// sha256 hex of the raw YAML block bytes, fences excluded.
    pub front_matter_sha: String,
    pub body: String,
    /// Every `[^cN]` in the body, in order, excluding footnote definitions.
    #[serde(default)]
    pub markers: Vec<String>,
    #[serde(default)]
    pub entry_errors: Vec<EntryError>,
}
